src/lucky_cli/task.cr

```diff
@@ -60,6 +68,131 @@ abstract class LuckyCli::Task
     end
   end
 
+  macro positional_arg(arg_name, description, required = nil, format = nil, to = nil)
+    {% PARSER_OPTS << arg_name %}
+    @{{ arg_name.id }} : {% if to %}Array(String){% else %}String{% end %} | Nil
+
+    def set_opt_for_{{ arg_name.id }}(args : Array(String))
+      {% if to %}
+        value = args[@positional_arg_count..-1]
+      {% else %}
+        value = args[@positional_arg_count]?
+      {% end %}
+      {% if format %}
+      matches = value.is_a?(Array) ? value.all?(&.=~({{ format }})) : value =~ {{ format }}
+      if matches
+        @{{ arg_name.id }} = value
+      else
+        raise "Invalid format for {{ arg_name.id }}. It should match {{ format }}"
+      end
+      {% else %}
+        @{{ arg_name.id }} = value
+      {% end %}
+      @positional_arg_count += 1
+    end
+
+    def {{ arg_name.id }}
+      {% if required %}
+        if @{{ arg_name.id }}.nil?
+          raise <<-ERROR
+          {{ arg_name.id }} is required, but no value was passed.
+          ERROR
+        end
+        @{{ arg_name.id }}.not_nil!
+      {% else %}
+        @{{ arg_name.id }}
+      {% end %}
+    end
```

Comment:
Do you think `required` should be checked in `set_opt_for_...`? That makes a little more sense to me than waiting until you try to access the method. This way you would not be allowing bad data in, and it would consolidate the validations (format and required).
```suggestion
    def set_opt_for_{{ arg_name.id }}(args : Array(String))
      {% if to_end %}
        value = args[@positional_arg_count..-1]
      {% else %}
        value = args[@positional_arg_count]?
      {% end %}
      {% if required %}
        if value.nil?
          raise <<-ERROR
          {{ arg_name.id }} is required, but no value was passed.
          ERROR
        end
      {% end %}
      {% if format %}
      matches = value.is_a?(Array) ? value.all?(&.=~({{ format }})) : value =~ {{ format }}
      if !matches
        raise "Invalid format for {{ arg_name.id }}. It should match {{ format }}"
      end
      {% end %}
      @{{ arg_name.id }} = value
      @positional_arg_count += 1
    end
    def {{ arg_name.id }}
      @{{ arg_name.id }}{% if required %}.not_nil!{% end %}
    end
```

Reply:
I agree, it would be better to check sooner. On positional args this is possible, but on flagged args, it's a little more difficult. I wanted it to be consistent so you weren't guessing when it would raise an error and when it wouldn't. I left a comment on the other one about the issue I ran in to, so if you have an idea for how to get around that, let me know.